datafusion/core/src/physical_optimizer/pipeline_checker.rs

```diff
@@ -0,0 +1,1173 @@
+///    Left   | Finite       |    false     | Hash         |\false
+///           | Data source  |--------------| Repartition  | \   +--------------+       +--------------+
+///           |              |              |              |  \  |              | true  |              | true
+///           +--------------+              +--------------+   - |  Hash Join   |-------| Projection   |-----
+///                                                            - |              |       |              |
+///           +--------------+              +--------------+  /  +--------------+       +--------------+
+///           |              |  unbounded   |              | /
+///    Right  | Infinite     |    true      | Hash         |/true
+///           | Data Source  |--------------| Repartition  |
+///           |              |              |              |
+///           +--------------+              +--------------+
+///
+/// ```
+fn hash_join_swap_subrule(
+    input: &PipelineStatePropagator,
+) -> Option<Result<PipelineStatePropagator>> {
+    let plan = input.plan.clone();
+    let children = &input.children_unbounded;
+    if let Some(hash_join) = plan.as_any().downcast_ref::<HashJoinExec>() {
+        let (left_unbounded, right_unbounded) = (children[0], children[1]);
+        let new_plan = match (left_unbounded, right_unbounded) {
+            (true, false) => {
+                if matches!(
+                    *hash_join.join_type(),
+                    JoinType::Inner
+                        | JoinType::Left
+                        | JoinType::LeftSemi
+                        | JoinType::LeftAnti
+                ) {
+                    swap(hash_join)
+                } else {
+                    Ok(plan)
+                }
+            }
+            _ => Ok(plan),
+        };
+        match new_plan {
+            Ok(plan) => Some(Ok(PipelineStatePropagator {
+                plan,
+                unbounded: left_unbounded || right_unbounded,
+                children_unbounded: vec![left_unbounded, right_unbounded],
+            })),
+            Err(e) => Some(Err(e)),
+        }
+    } else {
+        None
+    }
+}
+
+/// This function swaps sides of a hash join to make it runnable even if one of its
+/// inputs are infinite. Note that this is not always possible; i.e. [JoinType::Full],
+/// [JoinType::Left], [JoinType::LeftAnti] and [JoinType::LeftSemi] can not run with
+/// an unbounded left side, even if we swap. Therefore, we do not consider them here.
+fn swap(hash_join: &HashJoinExec) -> Result<Arc<dyn ExecutionPlan>> {
+    let partition_mode = hash_join.partition_mode();
+    let join_type = hash_join.join_type();
+    match (*partition_mode, *join_type) {
+        (
+            _,
+            JoinType::Right | JoinType::RightSemi | JoinType::RightAnti | JoinType::Full,
+        ) => Err(DataFusionError::Internal(format!(
+            "{} join cannot be swapped.",
```

Comment:
```suggestion
            "{} join cannot be swapped for unbounded input.",
```

Reply:
I recommend making these messages more specific to aid `grep`ing for them if they are hit in end to end tests